docs: make the docs more first-user-friendly (#695)
timescale/pgai: docs/vectorizer/README.md

# Automate AI embedding with pgai Vectorizer

Vector embeddings transform text and documents into compact, semantically rich representations that enable context-aware searches beyond traditional keyword matching.

While PostgreSQL excels at storing and querying these embeddings, synchronizing embeddings with source data has traditionally required custom solutions from developers.

> pgai Vectorizer provides a SQL-level interface that automates the embedding process within your database. By treating embeddings as a declarative feature—similar to an index—we simplify the entire workflow.

Reviewer: don't love "SQL-level" as it's a python lib...

Author: I still feel we are continuously talking about different things when using the word Vectorizer. This comes from https://github.com/timescale/pgai/pull/695/files#diff-918dd3e75368bf16064b77ffdb8a33ebfee6d8fcf268f076732e4b77fcf6a4ebL14-L25. 

We sometimes refer to the Vectorizer as a whole solution which includes pgai library + Vectorizer worker. Sometimes we just refer to the actual config model found in `ai.vectorizer`.

With pgai Vectorizer, you can:

- Designate text columns or documents for embedding using customizable rules
- Automatically generate and maintain searchable embedding tables 
- Keep embeddings synchronized with source data (asynchronously)
- Access a view that joins base tables with their embeddings

This page provides an overview of Vectorizer features. For quick setup instructions, see the 
[Vectorizer quick start](/docs/vectorizer/quick-start.md). For technical details, see the
[Vectorizer API reference](/docs/vectorizer/api-reference.md).

> Vectorizer uses a background worker for embedding generation, making the process performant and resilient to LLM endpoint failures. On [Timescale Cloud](https://tsdb.co/gh-pgai-signup), the worker runs automatically. For other cloud providers (AWS RDS, Supabase, etc.) or self-hosted Postgres, use the [vectorizer worker](/docs/vectorizer/worker.md) to process your vectorizers.

Reviewer: ```suggestion
Vectorizer uses a background worker for embedding generation, making the process performant and resilient to LLM endpoint failures. On [Timescale Cloud](https://tsdb.co/gh-pgai-signup), the worker runs automatically. For other cloud providers (AWS RDS, Supabase, etc.) or self-hosted Postgres, run the [vectorizer worker](/docs/vectorizer/worker.md) to process your vectorizers.
```

## How vectorizer works

> The vectorizer is designed to be flexible and customizable. Each vectorizer defines a pipeline for creating embeddings from your data. The pipeline consists of these components:

Reviewer: at this point people haven't even seen a create_vectorizer statement so does this make sense to them?

> - **[Loading](/docs/vectorizer/api-reference.md#loading-configuration):** Defines the source of data to embed - either directly from a database column or from an external file via URI

Reviewer: ```suggestion
- **[Loading](/docs/vectorizer/api-reference.md#loading-configuration):** Defines the source of data to embed - either directly from a database column or from external storage via URI to a file, AWS s3 bucket, etc.
```

Author: I think I'm gonna literally copy what you added in the main README.md: https://github.com/timescale/pgai/blob/main/README.md?plain=1#L335-L339

WDYT?

- **[Parsing](/docs/vectorizer/api-reference.md#parsing-configuration):** Converts non-text documents (PDF, HTML, markdown) into a text format suitable for embedding
- **[Chunking](/docs/vectorizer/api-reference.md#chunking-configuration):** Splits text data into smaller, semantically meaningful pieces
- **[Formatting](/docs/vectorizer/api-reference.md#formatting-configuration):** Prepares each chunk for embedding, optionally adding context like document title or metadata
- **[Embedding](/docs/vectorizer/api-reference.md#embedding-configuration):** Generates vector embeddings using your chosen LLM provider and model

Let's explore how the Vectorizer can transform your approach to unstructured data analysis and semantic search:

> - [Select an embedding provider and set up your API Keys](#select-an-embedding-provider-and-set-up-your-api-keys)

Reviewer: I think this table of contents may need to be updated. It's missing the appendix.

Author: I believe it can be even removed, GH already provides such a feature. Otherwise, if we still wanna include a TOC, should be generated automatically.

| chunk_seq      | INT    | Sequence number of the chunk within the document, starting at 0 |

To find the closest embeddings to a query, use this canonical SQL query:

```sql
SELECT 
   chunk,
   embedding <=> <query embedding> as distance
FROM blog_contents_embeddings
ORDER BY distance
LIMIT 10;
```

 The `<=>` operator calculates the distance between the query embedding and each
row's embedding vector. This is a simple way to do semantic search.

**Tip**: You can use the `ai.ollama_embed` function in our [PostgreSQL extension](/projects/extension/README.md) to generate an embedding for a user-provided query right inside the database.

You can combine this with metadata filters by adding a WHERE clause:

```sql
SELECT 
   chunk,
   embedding <=> <query embedding> as distance
FROM blog_contents_embeddings
WHERE 
   metadata->>'department' = 'finance'
ORDER BY 
   distance 
LIMIT 10;
```

This approach works with any column from the original table. For example, to search by author:

```sql
SELECT 
   chunk,
   embedding <=> <query embedding> as distance,
   author
FROM blog_contents_embeddings       
WHERE 
   author = 'Bulgakov'    
ORDER BY 
   distance 
LIMIT 10;
```

> ### Advanced query patterns with document embeddings

Reviewer: ```suggestion
### Advanced query patterns
```

A scheduled job checks for work to be done and, if so, runs the cloud function to embed the data.
By default, scheduling uses TimescaleDB background jobs running every five minutes.
Once the table is large enough, scheduling also handles index creation on the embedding column. 

When you self-host vectorizer, the vectorizer worker uses a polling mechanism to check whether
there is work to be done. Thus, scheduling is not needed and is deactivated by default.

Note: when scheduling is disabled, the index is not created automatically. You need to create it manually.

## The embedding storage table

The view is based on a table storing embeddings, named
`blog_contents_embeddings_store`. You can query this table directly for
potentially more efficient queries. The table structure is as follows:

```sql
CREATE TABLE blog_contents_embeddings_store(
    embedding_uuid UUID NOT NULL PRIMARY KEY DEFAULT gen_random_uuid(),     
    id INT,  -- primary key referencing the blog table
    chunk_seq INT NOT NULL, 
    chunk TEXT NOT NULL,
    embedding VECTOR(768) NOT NULL,
    UNIQUE (id, chunk_seq),
    FOREIGN KEY (id) REFERENCES public.blog(id) ON DELETE CASCADE
);
```

> ## Destination Options for Embeddings

Reviewer: This only really applies to text column embeddings...
